### ISSUE TITLE
Implement quasi-random generators for JAX

### ISSUE THREAD

**Issue by AdrienCorenflos** [google/jax#8807]
Hi,

Feature description
-------------------------
Scipy recently introduced (randomised) QMC sampler in their [API](https://docs.scipy.org/doc/scipy/reference/stats.qmc.html). The feature proposed in this issue is to bring this type of samplers in JAX too. 

Resources available
---------------------------
There exist a number of sources for these sequences on CPU (scipy runs on Cython, then you also have the [SOBOL](https://people.sc.fsu.edu/~jburkardt/cpp_src/sobol/sobol.html) cpp library, etc), and a scrambled Sobol sequence is implemented in [CURAND](https://docs.nvidia.com/cuda/curand/device-api-overview.html#quasirandom-sequences) (maybe Halton too)

Motivating example
--------------------------
The Sobol sequence is used in sequential quasi Monte Carlo method (see [paper](https://rss.onlinelibrary.wiley.com/doi/full/10.1111/rssb.12104) and [an implementation](https://github.com/nchopin/particles/blob/d389e8a4983513652ea528177eeee1c122b61c5f/particles/core.py#L208)). This is the only non-existing piece that is not trivially parallelisable with respect to the number of samples required. Obtaining this would help in porting these algorithms to the GPU world. It would also allow to use them in recently proposed sequential Monte Carlo based ELBO losses (https://arxiv.org/abs/1705.10306, https://arxiv.org/abs/1705.09279, https://arxiv.org/abs/1705.11140).

WDYT?

**Comment by mattjj** [comment]
It's a cool idea! But maybe best implemented in a higher-level library on top of JAX?

**Comment by AdrienCorenflos** [comment]
I'm a bit naive when it comes to CUDA bindings, but how would the dependency management work in this case? It tends to feel like having it closer to the core library would be somewhat more convenient, especially as the core random library already lives withing JAX? 

Now that I think about it, the question of parallelisation is a bit tough too. E.g. what would the below mean?

```python
def fun(qmc_state):
    return sobol.normal(qmc_state, qmc_state, (100,))

qmc_states = jax.qmc.split(init_state, 5)  # ???
jax.vmap(fun)(qmc_states)
```
You can "skip" values in sequences, so that it's feasible to do `qmc.skip(state, 100)`, but then to be able to `vmap` anything you need to know the number of samples you will need a priori, so that it rules out while loops within parallel settings.
Maybe splitting could mean "change the direction numbers" somehow?

**Comment by mattjj** [comment]
This sounds like a cool idea to explore, but it's out of scope for JAX core, at least right now. If you investigate this (e.g. in a separate library), I'd be personally interested to hear your findings! But since it's out of scope I'm going to close the issue.

**Comment by AdrienCorenflos** [comment]
oki doki

**Comment by srvasude** [comment]
FWIW in TFP on JAX, we have `tfp.mcmc.sample_halton_sequence` if that works for you.

**Comment by luweizheng** [comment]
@AdrienCorenflos is there any new ideas or any higher libraries built on JAX now? I want to use sobol or halton sequences, too.

**Comment by AdrienCorenflos** [comment]
Hi,
I kind of halted the project that required it (scrambled Sobol sequences) as the theoretical part ended up being tougher than expected so I don't have anything to offer, sorry.

**Comment by AdrienCorenflos** [comment]
Also, re. Sobol sequences, the scrambling can probably be done in parallel across the samples, but the actual sequence will need to be generated on CPU, at least I think so. 
The scrambling may be a bit technically tough but not impossible.

**Comment by carlosgmartin** [comment]
Related: https://github.com/tensorflow/probability/issues/181#issuecomment-1501230599. As @srvasude pointed out, Tensorflow has [a JAX implementation](https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax/mcmc/sample_halton_sequence) of Halton sequences, but I don't see one for Sobol sequences (which perform better, to my knowledge).

Also, PyTorch has added a [SobolEngine](https://pytorch.org/docs/stable/generated/torch.quasirandom.SobolEngine.html).

**Comment by gil2rok** [comment]
+1 for this feature request

**Comment by carlosgmartin** [comment]
@mattjj I've created a fairly compact implementation of the [Roberts sequence](https://extremelearning.com.au/unreasonable-effectiveness-of-quasirandom-sequences/):

https://gist.github.com/carlosgmartin/1fd4e60bed526ec8ae076137ded6ebab

Here are the outputs:

![image](https://github.com/user-attachments/assets/fc2b9a2c-af81-43d2-bbe1-caec6412e1f6)

![image](https://github.com/user-attachments/assets/a43b0e7c-e4a7-434d-9613-f0dae8413240)

I'd be happy to contribute this implementation to [`jax.random`](https://jax.readthedocs.io/en/latest/jax.random.html).

**Comment by carlosgmartin** [comment]
@mattjj @jakevdp Is there interest in merging https://github.com/jax-ml/jax/pull/23808? Or should I close that PR?

(Personally, I think quasi-random sampling is very useful across many fields, and this implementation doesn't add a lot of code.)

**Comment by mattjj** [comment]
Thanks for the ping! I had missed that before. I’m on my phone at the moment but it looks like really nice code. I’ll try to take a look.

**Comment by theo-brown** [comment]
I'd be interested in this too. Would it be possible to re-open this issue?

**Comment by gil2rok** [comment]
Would be interested as well!
